faculty/clients/api.py

```diff
@@ -0,0 +1,745 @@
+# Copyright 2018-2021 Faculty Science Limited
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+Manage Faculty APIs.
+"""
+import uuid
+from datetime import datetime
+from enum import Enum
+from typing import List, Optional, Union
+
+from attr import define
+from marshmallow import fields, post_load
+from marshmallow_enum import EnumField
+
+from faculty._oneofschema import OneOfSchema
+from faculty.clients.base import BaseClient, BaseSchema
+from faculty.clients.job import InstanceSize, _InstanceSizeSchema
+from faculty.clients.server import Server, _ServerSchema
+
+
+class DeploymentStatus(Enum):
+    """An enumeration of possible API deployment statuses."""
+
+    NOTDEPLOYED = "not-deployed"
+    STARTING = "starting"
+    DEPLOYED = "deployed"
+    ERROR = "error"
+
+
+@define
+class DevInstance:
+    instance_id: uuid.UUID
+
+
+@define
+class APIKey:
+    id: uuid.UUID
+    material: str
+    enabled: bool
+    label: str
+
+
+@define
+class ServerType:
+    instance_size_type: str
+    instance_size: Optional[InstanceSize] = None
+
+
+@define
+class WSGIDefinition:
+    working_directory: str
+    module: str
+    wsgi_object: str
+    last_updated_at: Optional[datetime] = None
```

Comment:
From looking at the aperture code, I don't think this is `Optional` (see ControllerReads line 92 and PersistentDetails line 43).

Reply:
It is optional, because this class is used not just for replies from `aperture`, but also request for creating APIs, and then the user would need to supply a `last_updated_at` value.

ie. with the current `Optional` setting, we can do

```python
api_definition = WSGIDefinition(working_directory="/project/testapi", module="app", wsgi_object="app")
# [...snip...]
x = client.create(project_id=PROJECT_ID, command_definition=api_definition, subdomain=subdomain, name=name, default_server_size=server)
```
and without it the request will fail, obviously, unless a value is set, even if the `create` endpoint doesn't require that value (see the Command Definitions in the README).

Thus the tricky place is that for requests it's definitely optional, for replies it will always be included, and thus the reason why I did it like this.

Given all the above, how would you prefer to resolve it?

* keeping it `Optional` due to the dual use?
* making it non-optional but having a default value? (of `datetime.now()` or something?) (this feels dirty)
* making it non-optional and make people pass in some value? (feels dirty like above)
* something completely different?

Reply:
Hmm, the "right" thing to do here would be to have different models, since the schema for the create body is different than the schema for the get response. But that will cause a fair bit of duplication, which is obviously undesirable. I'll have a think and get back to you.

I'd say a definite no to the second and third bullets in your comment.

Reply:
Hey @acroz, I've pushed some changes to see if they resolve this conundrum, breaking classes up into "Something" (for requests) and "SomethingResponse" (with inheritance derived from Something, for responses). This _feels_ like having the right amount of duplication (ie. "SomethingResponse" only needs the additional or changed fields). I've tested it very briefly so far and seems to work, also the way we are creating the `json` payload, that's helpful that we don't need to convert Something <-> SomethingResponse at any point, and still can use them. (that conversion would be a bit verbose and hacky, as much as I've researched).

It's not fully fledged yet (ie. there might be mistakes in the rewrite), but wanted to get your opinion with code.